text/0000-c-unwind-abi.md

```diff
@@ -0,0 +1,475 @@
+will specify this in [a future RFC][unresolved-questions].
+
+[inside-rust-forced]: https://blog.rust-lang.org/inside-rust/2020/02/27/ffi-unwind-design-meeting.html#forced-unwinding
+
+## Changes to `extern "C"` behavior
+[extern-c-behavior]: #changes-to-extern-c-behavior
+
+Prior to this RFC, any unwinding operation that crossed an `extern "C"`
+boundary, either from a `panic!` "escaping" from a Rust function defined with
+`extern "C"` or by entering Rust from another language via an entrypoint
+declared with `extern "C"`, caused undefined behavior.
+
+This RFC retains most of that undefined behavior, with one exception: with the
+`panic=unwind` runtime, `panic!` will cause an `abort` if it would otherwise
+"escape" from a function defined with `extern "C"`.
+
+## Interaction with `panic=abort`
+
+If a non-forced foreign unwind would enter a Rust frame via an `extern "C
+unwind"` ABI boundary, but the Rust code is compiled with `panic=abort`, the
+unwind will be caught and the process aborted.
+
+With the exception of the above case, however, unwinding from another language
+into Rust through an FFI entrypoint declared with `extern "C"` is always
+undefined behavior, and is not guaranteed to cause the program to abort under
+`panic=abort`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This table shows the behavior of an unwinding operation reaching each type of
+ABI boundary (function declaration or definition). "UB" stands for undefined
+behavior. `"C"`-like ABIs are `"C"` itself but also related ABIs such as
+`"system"`.
+
+| panic runtime  | ABI          | `panic`-unwind                        | Unforced foreign unwind |
+| -------------- | ------------ | ------------------------------------- | ----------------------- |
+| `panic=unwind` | `"C unwind"` | unwind                                | unwind                  |
+| `panic=unwind` | `"C"`-like   | abort                                 | UB                      |
+| `panic=abort`  | `"C unwind"` | `panic!` aborts                       | abort                   |
+| `panic=abort`  | `"C"`-like   | `panic!` aborts (no unwinding occurs) | UB                      |
+
+The interaction of Rust frames with C functions that deallocate frames (i.e.
+functions that may use forced unwinding on specific platforms) is independent
+of the panic runtime, ABI, or platform.
+
+* **When deallocating Rust non-POFs:** this is explicitly undefined behavior.
+* **When deallocating Rust [POFs][POF-definition]:** for now, this is not
+  specified, and must be considered undefined behavior. However, we do plan to
+  specify a safe way to deallocate POFs with `longjmp` or `pthread_exit` in [a
+  future RFC][unresolved-questions].
+
+No subtype relationship is defined between functions or function pointers using
+different ABIs. This RFC also does not define coercions between `"C"` and
+`"C unwind"`.
```

Comment:
Is that for reasons of scope or of soundness?

Reply:
Scope, I think. I actually am not sure that's the right call. I'd prefer to avoid a subtyping relationship, but adding coercions seems easy and harmless enough.

Reply:
It does imply that the "unwind" variant of an ABI is not actually *different* than the "ordinary" one in particular. This is true in all cases AFAIK. (i.e., three is really only one ABI underneath, and we are creating two ABIs indicating that you will not make use of certain capabilities)

Reply:
Subtyping implies that (but I agree we shouldn't do subtyping). Coercions could actually introduce a shim to convert between ABIs.

Reply:
Yes, sorry I was unclear. In any case I think that constraint is true. I just don't want to extend subtyping for impl reasons and because I don't think it's that important.

Reply:
I believe shims are still possible even without coercions:

```
extern "C" { fn foo(); }

fn extern "C unwind" foo_unwind() {
    foo()
}
```

...that's fairly verbose, though. Did you mean that coercions would make such shims unnecessary?

Reply:
What I meant is, subtyping can't introduce shims, so it would have to be true ABI compatibility if we wanted to do subtyping. But we actually do not want that.

Whether or not we coerce is entirely independent of whether they are ABI compatible. The latter should be specified though as it determines whether people are allowed to cast/transmute the fn ptr types around.

Reply:
That's a good point. I think we should specify that it is possible to transmute unless we know of at least one instance where that would be not be true.

Reply:
Is there something that I need to change in the draft?

Reply:
The text should say whether "this RFC also does not define coercions" because such coercions would be unsound, or to keep the scope of the RFC smaller.